internal/pkg/service/appproxy/http/router_test.go

```diff
@@ -342,30 +292,203 @@ func TestAppProxyRouter(t *testing.T) {
 
 				// Request to private app
 				rec = httptest.NewRecorder()
-				req = httptest.NewRequest("GET", "https://oidc.data-apps.keboola.local/", nil)
+				req = httptest.NewRequest(http.MethodGet, "https://oidc.data-apps.keboola.local/", nil)
+				for _, cookie := range cookies {
+					req.Header.Add("Cookie", cookie)
+				}
+				handler.ServeHTTP(rec, req)
+				require.Equal(t, http.StatusFound, rec.Code)
+			},
+		},
+		{
+			name: "private-app-oidc-down",
+			run: func(t *testing.T, handler http.Handler, m *mockoidc.MockOIDC, appServer *appServer) {
+				m.Shutdown()
+
+				client := createNoRedirectHTTPClient()
+
+				// Request to private app (unauthorized)
+				rec := httptest.NewRecorder()
+				req := httptest.NewRequest(http.MethodGet, "https://oidc.data-apps.keboola.local/", nil)
+				handler.ServeHTTP(rec, req)
+				require.Equal(t, http.StatusFound, rec.Code)
+				location := rec.Header()["Location"][0]
+				cookies := rec.Header()["Set-Cookie"]
+				assert.Len(t, cookies, 1)
+				wildcards.Assert(t, "_oauth2_proxy_csrf=%s; Path=/; Domain=oidc.data-apps.keboola.local; Expires=%s; HttpOnly; Secure", cookies[0])
+
+				// Request to the OIDC provider
+				request, err := http.NewRequestWithContext(context.Background(), http.MethodGet, location, nil)
+				require.NoError(t, err)
+				response, err := client.Do(request)
+				require.Error(t, err)
+				require.Nil(t, response)
+
+				// Request to private app (unauthorized)
+				rec = httptest.NewRecorder()
+				req = httptest.NewRequest(http.MethodGet, "https://oidc.data-apps.keboola.local/", nil)
+				for _, cookie := range cookies {
+					req.Header.Add("Cookie", cookie)
+				}
+				handler.ServeHTTP(rec, req)
+				require.Equal(t, http.StatusFound, rec.Code)
+			},
+		},
+		{
+			name: "private-app-down",
+			run: func(t *testing.T, handler http.Handler, m *mockoidc.MockOIDC, appServer *appServer) {
+				appServer.Close()
+
+				m.QueueUser(&mockoidcCustom.MockUser{
+					Email:  "admin@keboola.com",
+					Groups: []string{"admin"},
+				})
+
+				client := createNoRedirectHTTPClient()
+
+				// Request to private app (unauthorized)
+				rec := httptest.NewRecorder()
+				req := httptest.NewRequest(http.MethodGet, "https://oidc.data-apps.keboola.local/", nil)
+				handler.ServeHTTP(rec, req)
+				require.Equal(t, http.StatusFound, rec.Code)
+				location := rec.Header()["Location"][0]
+				cookies := rec.Header()["Set-Cookie"]
+				assert.Len(t, cookies, 1)
+				wildcards.Assert(t, "_oauth2_proxy_csrf=%s; Path=/; Domain=oidc.data-apps.keboola.local; Expires=%s; HttpOnly; Secure", cookies[0])
+
+				// Request to the OIDC provider
+				request, err := http.NewRequestWithContext(context.Background(), http.MethodGet, location, nil)
+				require.NoError(t, err)
+				response, err := client.Do(request)
+				require.NoError(t, err)
+				require.Equal(t, http.StatusFound, response.StatusCode)
+				location = response.Header["Location"][0]
```

Comment:
these parts of the code are repeated. Maybe you can create a function for that, but it's a detail.